servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RetryingHttpRequesterFilter.java

```diff
@@ -17,56 +17,134 @@
 import java.time.Duration;
+import java.util.function.BiFunction;
 import java.util.function.BiPredicate;
+import java.util.function.Function;
+import javax.annotation.Nullable;
 
+import static io.servicetalk.concurrent.api.AsyncCloseables.emptyAsyncCloseable;
+import static io.servicetalk.concurrent.api.AsyncCloseables.toAsyncCloseable;
+import static io.servicetalk.concurrent.api.Completable.completed;
 import static io.servicetalk.concurrent.api.Completable.failed;
+import static io.servicetalk.concurrent.api.RetryStrategies.retryWithConstantBackoffDeltaJitter;
+import static io.servicetalk.concurrent.api.RetryStrategies.retryWithConstantBackoffFullJitter;
+import static io.servicetalk.concurrent.api.RetryStrategies.retryWithExponentialBackoffDeltaJitter;
+import static io.servicetalk.concurrent.api.RetryStrategies.retryWithExponentialBackoffFullJitter;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HeaderUtils.DEFAULT_HEADER_FILTER;
+import static java.time.Duration.ZERO;
+import static java.time.Duration.ofDays;
+import static java.util.Objects.requireNonNull;
 
 /**
  * A filter to enable retries for HTTP requests.
- *
+ * <p>
+ * Retries are supported for both the request flow and the response flow. Retries, in other words, can be triggered
+ * as part of a service response if needed.
+ * <p>
+ * The two behaviors can have different criteria, as defined from the relevant Builder methods (i.e.
+ * {@link RetryingHttpRequesterFilterBuilder#retryRequests(BiFunction)}
+ * or {@link RetryingHttpRequesterFilterBuilder#retryResponses(Function)}).
+ * Both return a {@link BackOffPolicy} when the request or the response is matching the conditions, which allows
+ * control of the retry backoff period independently.
+ * Similarly, max-retries for each flow can be set in the {@link BackOffPolicy}, as well
+ * as a total max-retries to be respected by both flows, as set in
+ * {@link RetryingHttpRequesterFilterBuilder#maxTotalRetries(int)}.
+ * <p>
+ * If the existing preset of {@link BackOffPolicy backoff policies} aren't enough to meet your expectations, the
+ * class is extendable.
+ * <p>
+ * Note that applying this filter on a client it will automatically disable the use of
+ * {@link SingleAddressHttpClientBuilder#appendClientFilter(StreamingHttpClientFilterFactory)}.
  * @see RetryStrategies
  */
-public final class RetryingHttpRequesterFilter implements StreamingHttpClientFilterFactory,
-                                                          StreamingHttpConnectionFilterFactory,
-                                                          HttpExecutionStrategyInfluencer {
+public final class RetryingHttpRequesterFilter
+        implements StreamingHttpClientFilterFactory, StreamingHttpConnectionFilterFactory,
+                   HttpExecutionStrategyInfluencer, InjectableStreamingClientFilterFactory,
+                   LoadBalancerReadinessAware, ServiceDiscoveryStatusAware {
```

Comment:
Main issues with the approach proposed in this PR:
1. There are no retries by default, which breaks the client for all users because it has to wait for LB to be ready. 
2. If a filter is applied as `StreamingHttpConnectionFilterFactory`, there are multiple issues:
   a. wait for LB and SD error visibility won't work because we can not create a connection without having LB ready.
   b. `RetryableException` has to be handled at a client level, because LB may throw it.
3. SD and LB are protocol agnostic abstractions and expected to be reused outside HTTP. This is why they are defined in client-api. This is why `AutoRetryStrategyProvider` also lives in client-api.
4. `InjectableStreamingClientFilterFactory`, `LoadBalancerReadinessAware`, and `ServiceDiscoveryStatusAware` interfaces look like hacks in API to workaround a specific case. I don't see any other cases when this API will be useful for users and how they will discover that these hooks exist. Can you provide examples?

Reply:
Discussed offline.

1. By default auto-retries are enabled (if they weren't disabled), but using the new filter. If the user provides a custom Retrying Filter, then auto-retry will be skipped.
2. Decided to remove support for connection based filters. Code was adjusted.
3. Decided to merge this together since there is no immediate need for supporting multiple protocols.
4. Agreed, removed.